node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -633,200 +630,149 @@ impl Initialized {
+			Some(votes) => VoteState::new(votes, &env),
 			None =>
 				if let MaybeCandidateReceipt::Provides(candidate_receipt) = candidate_receipt {
-					(
-						CandidateVotes {
-							candidate_receipt,
-							valid: Vec::new(),
-							invalid: Vec::new(),
-						},
-						true,
-					)
+					VoteState::new_from_receipt(candidate_receipt)
 				} else {
 					gum::warn!(
 						target: LOG_TARGET,
 						session,
-						"Not seen backing vote for candidate which has an active dispute",
+						"Cannot import votes, without `CandidateReceipt` available!"
 					);
 					return Ok(ImportStatementsResult::InvalidImport)
 				},
 		};
-		let candidate_receipt = votes.candidate_receipt.clone();
-		let was_concluded_valid = votes.valid.len() >= supermajority_threshold;
-		let was_concluded_invalid = votes.invalid.len() >= supermajority_threshold;
-
-		let mut recent_disputes = overlay_db.load_recent_disputes()?.unwrap_or_default();
-		let controlled_indices = find_controlled_validator_indices(&self.keystore, &validators);
-
-		// Whether we already cast a vote in that dispute:
-		let voted_already = {
-			let mut our_votes = votes.voted_indices();
-			our_votes.retain(|index| controlled_indices.contains(index));
-			!our_votes.is_empty()
-		};
 
-		let was_confirmed = recent_disputes
-			.get(&(session, candidate_hash))
-			.map_or(false, |s| s.is_confirmed_concluded());
+		let import_result = {
+			let intermediate_result = old_state.import_statements(&env, statements);
 
-		let is_included = self.scraper.is_candidate_included(&candidate_receipt.hash());
-
-		let is_local = statements
-			.iter()
-			.find(|(_, index)| controlled_indices.contains(index))
-			.is_some();
-
-		// Indexes of the validators issued 'invalid' statements. Will be used to populate spam slots.
-		let mut fresh_invalid_statement_issuers = Vec::new();
-
-		// Update candidate votes.
-		for (statement, val_index) in &statements {
-			if validators
-				.get(val_index.0 as usize)
-				.map_or(true, |v| v != statement.validator_public())
+			// Handle approval vote import:
+			//
+			// See guide: We import on fresh disputes to maximize likelihood of fetching votes for
+			// dead forks and once concluded to maximize time for approval votes to trickle in.
+			if intermediate_result.is_freshly_disputed() ||
+				intermediate_result.is_freshly_concluded()
 			{
-				gum::debug!(
-				target: LOG_TARGET,
-				?val_index,
-				session,
-				claimed_key = ?statement.validator_public(),
-				"Validator index doesn't match claimed key",
+				let (tx, rx) = oneshot::channel();
+				// Bounded because:
+				// 1. Only triggered twice per dispute.
+				// 2. Raising a dispute is costly (requires validation + recovery) by honest nodes,
+				// dishonest nodes are limited by spam slots.
+				// 3. Concluding a dispute is even more costly.
+				// Therefore it is reasonable to expect a simple vote request to succeed way faster
+				// than disputes are raised.
+				ctx.send_unbounded_message(
```

Comment:
I assume we want to prioritize this one over the unbounded channel, and we have guarantees that this is actually bounded. From what I see this could be a heavy operation.

Reply:
I meant the usage - clarified.